File: ICPC/2022-2023 ACM-ICPC Latin American Regional Programming Contest/A. Asking for Money.cpp
#include <bits/stdc++.h>
using namespace std;

const int MAXN = 1005;
vector<int> graph[MAXN];
int visited[MAXN];

void dfs(int node, int ancestor)
{
    visited[node] = 1;
    for (int neighbor : graph[node])
    {
        if (neighbor != ancestor && !visited[neighbor])
        {
            dfs(neighbor, ancestor);
        }
    }
}

int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    int n;
    cin >> n;
    vector<pair<int, int>> edges(n);
    for (int i = 0; i < n; ++i)
    {
        int u, v;
        cin >> u >> v;
        u--;
        v--;
        edges[i] = {u, v};
        graph[u].push_back(i);
        graph[v].push_back(i);
    }

    for (int i = 0; i < n; ++i)
    {
        vector<int> count(n);
        for (int node : {i, edges[i].first, edges[i].second})
        {
            memset(visited, 0, sizeof(visited));
            dfs(node, i);
            for (int j = 0; j < n; ++j)
            {
                count[j] += visited[j];
            }
        }

        int can = 0;
        for (int j = 0; j < n; ++j)
        {
            if (count[j] == 3)
            {
                can = 1;
                break;
            }
        }
        cout << (can ? "Y" : "N");
    }
    cout << "\n";

    return 0;
}

File: ICPC/2022-2023 ACM-ICPC Latin American Regional Programming Contest/A.Asking for Money.py

from collections import defaultdict


def dfs(node, ancestor):
    visited[node] = 1
    for neighbor in graph[node]:
        if neighbor != ancestor and not visited[neighbor]:
            dfs(neighbor, ancestor)


n = int(input())
visited = [0] * 1005
graph = defaultdict(list)
edges = []
for _ in range(n):
    u, v = map(int, input().split())
    u -= 1
    v -= 1
    edges.append((u, v))
    graph[u].append(_)
    graph[v].append(_)

for i in range(n):
    count = [0] * n
    for node in [i, edges[i][0], edges[i][1]]:
        visited = [0] * 1005
        dfs(node, i)
        for j in range(n):
            count[j] += visited[j]

    can = 0
    print(count)
    for j in range(n):
        if count[j] == 3:
            can = 1
            break
    # print("Y" if can else "N", end="")
